### ISSUE TITLE
Hang: Second request to same URL immediately after first

### ISSUE THREAD

**Issue by scottnonnenberg** [electron/electron#10570]
* Electron version: 1.8.0, 1.6.13, and 1.6.7
* Operating system: Ubuntu 16.04 (no repro on OSX 10.11.6)

### Expected behavior

Two repeated requests to the same URL from the renderer process via Node.js APIs come back with responses. On OSX this happens properly.

### Actual behavior

On Linux, if a second request is made right after the first one finishes, that second request will start but not finish. Strangely, it can be forced to come back later with another request.

### How to reproduce

Put this code in your `preload.js` file and start Electron:

```javascript
var https = require('https');

var _setImmediate = setImmediate;
process.once('loaded', function() {
  window.setImmediate = _setImmediate;
});

function makeRequest(cb) {
  if (!cb) {
    cb = function() {}
  }

  console.log('making request');
  var options = {
    host: 'github.com',
    port: 443,
    path: '/',
    method: 'GET',
    agent: new https.Agent()
  };
  var req = https.request(options, (res) => {
    console.log('request/statusCode:', res.statusCode);
    console.log('request/headers:', res.headers);

    res.on('data', (d) => {
      console.log('data', d.toString().length);
    });
    res.on('end', () => {
      console.log('No more data in response.');
      return cb();
    });
  });

  req.on('request/error', (e) => {
    console.error(e);
    return cb();
  });
  req.end();
}


makeRequest(function() {
  makeRequest();
});

if (typeof window !== 'undefined') {
  window.makeRequest = makeRequest;
}
```

On Linux the final log entry is `making request` because the second request hangs. On Electron 1.8.0, the DevTools disconnect a bit later, and you get a 'Check failed' log to the console (see the details below). On previous versions of Electron, neither of those two things happen.

You can call `window.makeRequest()` later and see two responses come back. This appears to 'wake up' the Node.js side.

### Other things tried

- Direct call of this code from Node.js 7.4.0 on Linux and OSX. No repro.
- Direct load from Electron entrypoint (the one that calls `createWindow`). No repro.
- Load of this code via `require()` and `remote.require()` from `preload.js`. Still happened.
- Looked at the logging that came from `NODE_DEBUG=http`, and key events are not fired when this issue repros, even at that level
- Tried it on windows both in Electron and in Node.js. No repro.

<Details>
The check failed log entry on Electron 1.8.0:

```
[22179:0921/130510.974808:FATAL:partition_alloc.cc(934)] Check failed: page->num_allocated_slots != -1. 
#0 0x000001deb897 <unknown>
#1 0x000001e163ab <unknown>
#2 0x000001dcfeb5 <unknown>
#3 0x000001d3045d <unknown>
#4 0x7f31a7178b43 <unknown>
#5 0x7f31a70ebeb2 <unknown>
#6 0x7f31a70f7a71 <unknown>
#7 0x7f31a70eccfe <unknown>
#8 0x7f31a70e5657 <unknown>
#9 0x7f31a70e3059 <unknown>
#10 0x7f31a6bc6523 <unknown>
#11 0x7f31a6bc4b7a <unknown>
#12 0x7f31a6bc40f4 <unknown>
#13 0x7f31a6bd2996 <unknown>
#14 0x7f31a6ef9fb3 <unknown>
#15 0x000001e03660 <unknown>
#16 0x000001c8cef8 <unknown>
#17 0x000001c8b198 <unknown>
#18 0x000001e03660 <unknown>
#19 0x000001df688d <unknown>
#20 0x000001df6ba8 <unknown>
#21 0x000001df7246 <unknown>
#22 0x000001e080ea <unknown>
#23 0x000001df65b7 <unknown>
#24 0x000001e5476e <unknown>
#25 0x00000302d350 <unknown>
#26 0x0000031721aa <unknown>
#27 0x000003173048 <unknown>
#28 0x0000029e111f <unknown>
#29 0x000003171ec2 <unknown>
#30 0x000003de1168 main
#31 0x7f319f94b830 __libc_start_main
#32 0x0000005a5099 <unknown>
```

DevTools disconnect on Electron 1.8.0:
![screen shot 2017-09-21 at 10 07 57 am](https://user-images.githubusercontent.com/443005/30709557-85716f5a-9eb7-11e7-80b4-2ff93a4583b6.png)
</Details>

**Comment by scottnonnenberg** [comment]
You might be wondering why we're trying to use Node.js APIs for making web requests in the renderer process - it's because we can provide our own certificate authority.

I'm really not sure where to go next. I was thinking that we could move all of our web requests into the main process and rely on IPC to get that data to the renderer process, but that's a pretty big change for our app. Any thoughts?

**Comment by scottnonnenberg** [comment]
The plot thickens:

I tested this on Windows 10, and it doesn't happen.

My repository for testing it in a pristine state does **NOT** show the behavior: https://github.com/scottnonnenberg/electron-hang-repro

What does show the behavior is the electron app here: https://github.com/WhisperSystems/Signal-Desktop/tree/v1.0.25

Next on my list: figure out the difference between which causes this.

**Comment by scottnonnenberg** [comment]
It turns out that the hang is intermittent in the small repository, where it's 100% likely in our full Signal Desktop repository.

Anyway, I've made some changes to the repro repository to help make it more likely. You can go uncomment some things in `blank.html` to make it load a bunch of browser javascript. https://github.com/scottnonnenberg/electron-hang-repro

In the full repository, as far as I can tell it's the [sheer amount of javascript files loaded and parsed](https://github.com/WhisperSystems/Signal-Desktop/blob/bd78016120d6a6312753b573ce7abe57b845cabe/background.html#L784-L850) which causes this to be 100% reproducible.

**Comment by refractalize** [comment]
* Electron version: 1.8.2, 1.8.4
* Operating system: Ubuntu 14.04.1

I'm seeing this too, only we're not making two identical requests, just one is enough to trigger it. We're making a HTTPS call, POST or GET, no matter the endpoint. It fails to make the HTTP request (no outbound request seen in a HTTP proxy), as though the `req.end()` is not being called - so no timeouts are fired. I can confirm @scottnonnenberg's fix works.

Was able to reproduce the bug in both the renderer process and the main process.

Was not able to reproduce the bug in the equivalent versions of node, nor was I able to reproduce on Mac OSX, or Linux Fedora.

**Comment by mcferden** [comment]
- Electron version: 1.8.4, 2.0.0
- Operating system: Ubuntu 17.10, Ubuntu 18.04

I have this issue too. Here is an example [repository](https://github.com/mcferden/electron-request-bug). The series of requests hangs after the first request completes. On my machine it reproduces every time I run the app.

This issue does not appear in pure Node.js and it is Linux specific, on Windows the app works fine.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by sofianguy** [comment]
Thank you for your issue!

We haven't gotten a response to our questions in our comment above. With only the information that is currently in the issue, we don't have enough information to take action. I'm going to close this but don't hesitate to reach out if you have or find the answers we need, we'll be happy to reopen the issue.

**Comment by dsvictor94** [comment]
I have the same problem, but it is hard to reproduce. Only happens in a specific Window 10 machine even after reinstall the system. But I couldn't reproduce in a VM with the same operation system.

@scottnonnenberg did you find something to fix the problem?

**Comment by scottnonnenberg** [comment]
This workaround seems to prevent the problem: https://github.com/signalapp/Signal-Desktop/blob/419f219f916619ac9292863e5049970a80390dd4/preload.js#L243-L246

**Comment by acouvreur** [comment]
We are having this issue too on linux (we didn't test yet on Windows). The workaround proposed by @scottnonnenberg seems to work for us.

This is a really weird behavior. You suppose that the event loop is stopped randomly, what made you think that?

We're using a hidden BrowserWindow as a server process to offload the Main process. 
When we log in, we directly load some data with a second request. And the second request will _sometimes_ wait between 10 to 20 seconds before getting sent.

However this behavior is random, and, as I said, sometimes **there is no waiting time**. We checked that our server was not faulty. We found that **the request was just not sent to our server before the 10 to 20 seconds delay**.

We also started our embedded server with Node, outside Electron, and we **could not reproduce this behavior**. We concluded that the issue was with electron BrowserWindow.

Let me know if you experience something similar. Thanks again @scottnonnenberg !
